Platform/OpenCanopy/Views/BootPicker.c

```diff
@@ -718,42 +726,180 @@ GLOBAL_REMOVE_IF_UNREFERENCED GUI_OBJ mBootPickerView = {
+
+  UseVolumeIcon   = (Context->ConsoleAttributes & OPENCANOPY_USE_VOLUME_ICON) != 0;
+  UseDiskLabel    = (Context->ConsoleAttributes & OPENCANOPY_USE_DISK_LABEL_FILE) != 0;
+  UseGenericLabel = (Context->ConsoleAttributes & OPENCANOPY_USE_GENERIC_LABEL_IMAGE) != 0;
+
+  DEBUG((DEBUG_INFO, "UseDiskLabel: %d, UseGenericLabel: %d\n", UseDiskLabel, UseGenericLabel));
 
   VolumeEntry = AllocateZeroPool (sizeof (*VolumeEntry));
   if (VolumeEntry == NULL) {
     return RETURN_OUT_OF_RESOURCES;
   }
 
-  Result = GuiGetLabel (
-             &VolumeEntry->Label,
-             &GuiContext->FontContext,
-             Name,
-             StrLen (Name)
-             );
-  if (!Result) {
-    DEBUG ((DEBUG_WARN, "BMF: label failed\n"));
-    return RETURN_UNSUPPORTED;
+  AppleBootPolicy = OcAppleBootPolicyInstallProtocol (FALSE);
+  if (AppleBootPolicy == NULL) {
+    DEBUG ((DEBUG_ERROR, "OCB: AppleBootPolicy locate failure\n"));
+    return EFI_NOT_FOUND;
+  }
+
+  if (UseDiskLabel
+   && EFI_SUCCESS == OcGetBootEntryLabelImage(Context, AppleBootPolicy, Entry, 1, &IconFileData, &IconFileSize)
+   && EFI_SUCCESS == DecodeAppleDiskLabelImage(&VolumeEntry->Label, IconFileData, IconFileSize)) {
+    // do nothing
+  } else if (UseGenericLabel) {
+    switch (Entry->Type) {
+      case OC_BOOT_UNKNOWN:
+      case OC_BOOT_EXTERNAL_OS:
+        Status = CopyLabel(&VolumeEntry->Label, &GuiContext->EntryLabelEFIBoot);
+        break;
+
+      case OC_BOOT_APPLE_RECOVERY:
+        Status = CopyLabel(&VolumeEntry->Label, &GuiContext->EntryLabelRecovery);
+        break;
+
+      case OC_BOOT_WINDOWS:
+        Status = CopyLabel(&VolumeEntry->Label, &GuiContext->EntryLabelWindows);
+        break;
+
+      case OC_BOOT_EXTERNAL_TOOL:
+        Status = CopyLabel(&VolumeEntry->Label, &GuiContext->EntryLabelTool);
+        break;
+
+      case OC_BOOT_APPLE_OS:
+        Status = CopyLabel(&VolumeEntry->Label, &GuiContext->EntryLabelMacOS);
+        break;
+
+      case OC_BOOT_SYSTEM:
+        if (StrCmp(Entry->Name, OC_MENU_RESET_NVRAM_ENTRY) == 0) {
+          Status = CopyLabel(&VolumeEntry->Label, &GuiContext->EntryLabelResetNVRAM);
+        } else {
+          Status = CopyLabel(&VolumeEntry->Label, &GuiContext->EntryLabelTool);
+        }
+        break;
+
+      default:
+        DEBUG((DEBUG_ERROR, "Entry kind %d unsupported", Entry->Type));
+        return RETURN_UNSUPPORTED;
+    }
+    if (RETURN_ERROR(Status)) {
+      return Status;
+    }
+  } else {
+    Result = GuiGetLabel (
+      &VolumeEntry->Label,
+      &GuiContext->FontContext,
+      Entry->Name,
+      StrLen (Entry->Name)
+      );
+    if (!Result) {
+      DEBUG ((DEBUG_WARN, "BMF: label failed\n"));
+      return RETURN_UNSUPPORTED;
+    }
   }
 
-  VolumeEntry->Context = EntryContext;
+  VolumeEntry->Context = Entry;
 
-  if (!IsExternal) {
+  if (UseVolumeIcon
+   && EFI_SUCCESS == OcGetBootEntryIcon(Context, AppleBootPolicy, Entry, &IconFileData, &IconFileSize)
```

Comment:
Where is IconFileData supposed to be freed?

Reply:
Hm, need to free it right here, but also in case of decode failure. Will extract it into a function so the if-else if-else control flow is still applicable